app/src/main/java/com/example/tasky/authentication/presentation/login/viewmodel/LoginViewModel.kt

```diff
@@ -1,14 +1,103 @@
 package com.example.tasky.authentication.presentation.login.viewmodel
 
-class LoginViewModel {
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.setValue
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.example.tasky.authentication.data.remote.LoginRepository
+import com.example.tasky.authentication.domain.usecase.LoginUseCase
+import com.example.tasky.core.data.Resource
+import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@HiltViewModel
+class LoginViewModel @Inject constructor(
+    private val loginUseCase: LoginUseCase,
+    private val loginRepository: LoginRepository
+) : ViewModel() {
+    var state by mutableStateOf(LoginState())
+        private set
+
+    private var loginJob: Job? = null
+
+    fun onEvent(event: LoginEvent) {
+        when (event) {
+            is LoginEvent.OnEmailValueChanged -> {
+                val isValid = loginUseCase.validateEmailUseCase(event.emailValue)
+                state = state.copy(isValidEmail = isValid, emailValue = event.emailValue)
+            }
+
+            is LoginEvent.OnPasswordValueChanged -> {
+                state = state.copy(passwordValue = event.passwordValue)
+            }
+
+            is LoginEvent.OnShowPasswordValueChanged -> {
+                state = state.copy(showPassword = !state.showPassword)
+            }
+
+            LoginEvent.OnLoginClicked -> {
+                if (!state.isValidEmail) {
+                    state = state.copy(errorMessage = "Invalid email")
+                    return
+                }
+
+                if (loginJob != null) return
+
+                state = state.copy(isLoading = true)
+                loginJob = viewModelScope.launch {
+                    val result = loginRepository.login(state.emailValue, state.passwordValue)
+
+                    state = when (result) {
+                        is Resource.Success -> {
+                            state.copy(
+                                onLoginSucceed = true,
+                                isLoading = false
+                            )
+                        }
+
+                        is Resource.Error -> {
+                            state.copy(
+                                errorMessage = result.errorMessage,
+                                isLoading = false
+                            )
+                        }
+                    }
+                    loginJob = null
+                }
+            }
+
+            LoginEvent.OnSignUpClicked -> {
+                state = state.copy(navigateToSignUp = true)
+            }
+        }
+    }
+
+    fun errorShown() {
```

Comment:
Why do you have these 3 public functions?

The whole point of having the LoginEvent is to just expose one single function, and act based on events

In this case, you can create 3 events (ErrorShown, LoginNavigated, SignupNavigated) and use the onEvent function


Also,
Since these 3 are to mimic a onetimeevent (restore the state)
Personally I think they add too much fluff. I prefer a Channel and just read it using LaunchedEffect
But I guess it works fine

Reply:
Yes, I personally prefer sticking to the UI -> Event -> Viewmodel and Viewmodel -> State -> UI model following [this](https://medium.com/androiddevelopers/viewmodel-one-off-event-antipatterns-16a1da869b95) and considering the advantages and disadvantages of each way

Reply:
Gotcha, sounds good